File: javascript/src/delegation.js
/**
 * Delegation attestation creation and verification for AgentPin.
 */

import { signData, verifySignature } from './crypto.js';
import { capabilitiesHash } from './capability.js';
import { AgentPinError, ErrorCode } from './types.js';

/**
 * Create the canonical attestation input string.
 * Format: `{domain}|{role}|{agent_id}|{delegatee_domain}|{delegatee_agent_id}|{capabilities_hash}`
 * @param {string} domain
 * @param {string} role
 * @param {string} agentId
 * @param {string} delegateeDomain
 * @param {string} delegateeAgentId
 * @param {Capability[]} capabilities
 * @returns {string}
 */
export function canonicalAttestationInput(domain, role, agentId, delegateeDomain, delegateeAgentId, capabilities) {
    const capHash = capabilitiesHash(capabilities);
    return `${domain}|${role}|${agentId}|${delegateeDomain}|${delegateeAgentId}|${capHash}`;
}

/**
 * Create a delegation attestation signed by the attesting entity.
 * @param {string} privateKeyPem
 * @param {string} kid
 * @param {string} domain
 * @param {string} role
 * @param {string} agentId
 * @param {string} delegateeDomain
 * @param {string} delegateeAgentId
 * @param {Capability[]} capabilities
 * @returns {object} DelegationAttestation
 */
export function createAttestation(
    privateKeyPem, kid, domain, role, agentId,
    delegateeDomain, delegateeAgentId, capabilities
) {
    const input = canonicalAttestationInput(
        domain, role, agentId, delegateeDomain, delegateeAgentId, capabilities
    );
    const signature = signData(privateKeyPem, Buffer.from(input));

    return {
        domain,
        role,
        agent_id: agentId,
        kid,
        attestation: signature,
    };
}

/**
 * Verify a single delegation attestation signature.
 * @param {object} attestation
 * @param {string} publicKeyPem
 * @param {string} delegateeDomain
 * @param {string} delegateeAgentId
 * @param {Capability[]} capabilities
 * @throws {AgentPinError}
 */
export function verifyAttestation(attestation, publicKeyPem, delegateeDomain, delegateeAgentId, capabilities) {
    const input = canonicalAttestationInput(
        attestation.domain, attestation.role, attestation.agent_id,
        delegateeDomain, delegateeAgentId, capabilities
    );

    const valid = verifySignature(publicKeyPem, Buffer.from(input), attestation.attestation);
    if (!valid) {
        throw new AgentPinError(
            ErrorCode.DELEGATION_INVALID,
            `Delegation attestation from ${attestation.domain} failed signature verification`
        );
    }
}

/**
 * Verify the depth of a delegation chain does not exceed the minimum max_delegation_depth.
 * @param {number} chainLen
 * @param {number[]} maxDepths
 * @throws {AgentPinError}
 */
export function verifyChainDepth(chainLen, maxDepths) {
    const minDepth = Math.min(...maxDepths);
    if (chainLen > minDepth) {
        throw new AgentPinError(
            ErrorCode.DELEGATION_DEPTH_EXCEEDED,
            `Delegation chain depth ${chainLen} exceeds minimum max_delegation_depth ${minDepth}`
        );
    }
}

File: python/agentpin/delegation.py
"""Delegation attestation creation and verification for AgentPin."""

from typing import List

from .capability import Capability, capabilities_hash
from .crypto import sign_data, verify_signature
from .types import AgentPinError, ErrorCode


def canonical_attestation_input(
    domain: str,
    role: str,
    agent_id: str,
    delegatee_domain: str,
    delegatee_agent_id: str,
    capabilities: List[Capability],
) -> str:
    """Create the canonical attestation input string.

    Format: {domain}|{role}|{agent_id}|{delegatee_domain}|{delegatee_agent_id}|{capabilities_hash}
    """
    cap_hash = capabilities_hash(capabilities)
    return f"{domain}|{role}|{agent_id}|{delegatee_domain}|{delegatee_agent_id}|{cap_hash}"


def create_attestation(
    private_key_pem: str,
    kid: str,
    domain: str,
    role: str,
    agent_id: str,
    delegatee_domain: str,
    delegatee_agent_id: str,
    capabilities: List[Capability],
) -> dict:
    """Create a delegation attestation signed by the attesting entity."""
    input_str = canonical_attestation_input(domain, role, agent_id, delegatee_domain, delegatee_agent_id, capabilities)
    signature = sign_data(private_key_pem, input_str.encode("utf-8"))

    return {
        "domain": domain,
        "role": role,
        "agent_id": agent_id,
        "kid": kid,
        "attestation": signature,
    }


def verify_attestation(
    attestation: dict,
    public_key_pem: str,
    delegatee_domain: str,
    delegatee_agent_id: str,
    capabilities: List[Capability],
) -> None:
    """Verify a single delegation attestation signature.

    Raises:
        AgentPinError: if signature verification fails.
    """
    input_str = canonical_attestation_input(
        attestation["domain"],
        attestation["role"],
        attestation["agent_id"],
        delegatee_domain,
        delegatee_agent_id,
        capabilities,
    )

    valid = verify_signature(public_key_pem, input_str.encode("utf-8"), attestation["attestation"])
    if not valid:
        raise AgentPinError(
            ErrorCode.DELEGATION_INVALID,
            f"Delegation attestation from {attestation['domain']} failed signature verification",
        )


def verify_chain_depth(chain_len: int, max_depths: List[int]) -> None:
    """Verify the depth of a delegation chain does not exceed the minimum max_delegation_depth.

    Raises:
        AgentPinError: if chain depth exceeds limits.
    """
    min_depth = min(max_depths) if max_depths else 0
    if chain_len > min_depth:
        raise AgentPinError(
            ErrorCode.DELEGATION_DEPTH_EXCEEDED,
            f"Delegation chain depth {chain_len} exceeds minimum max_delegation_depth {min_depth}",
        )
